File: core/core-sync/src/intTestDisabled/java/ch/difty/scipamato/core/sync/jobs/codeclass/CodeClassSyncConfigTest.java
package ch.difty.scipamato.core.sync.jobs.codeclass;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.*;

import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Timestamp;

import org.jooq.DeleteConditionStep;
import org.jooq.TableField;
import org.junit.jupiter.api.Test;
import org.mockito.Mockito;
import org.springframework.batch.core.Job;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;

import ch.difty.scipamato.core.db.tables.CodeClass;
import ch.difty.scipamato.core.db.tables.CodeClassTr;
import ch.difty.scipamato.core.sync.jobs.SyncConfigTest;
import ch.difty.scipamato.publ.db.tables.records.CodeClassRecord;

@SpringBootTest
class CodeClassSyncConfigTest extends SyncConfigTest<CodeClassRecord> {

    @Autowired
    private CodeClassSyncConfig config;

    @Override
    protected Job getJob() {
        return config.syncCodeClassJob();
    }

    @Override
    protected String selectSql() {
        return config.selectSql();
    }

    @Override
    protected TableField<CodeClassRecord, Timestamp> lastSynchedField() {
        return config.lastSynchedField();
    }

    @Override
    protected DeleteConditionStep<CodeClassRecord> getPseudoFkDcs() {
        return config.getPseudoFkDcs();
    }

    @Override
    protected String expectedJobName() {
        return "syncCodeClassJob";
    }

    @Override
    protected String expectedSelectSql() {
        return
            "select \"public\".\"code_class\".\"id\", \"public\".\"code_class_tr\".\"lang_code\", \"public\".\"code_class_tr\".\"name\", \"public\".\"code_class_tr\".\"description\", "
            + "\"public\".\"code_class_tr\".\"version\", \"public\".\"code_class_tr\".\"created\", \"public\".\"code_class_tr\".\"last_modified\" "
            + "from \"public\".\"code_class\" join \"public\".\"code_class_tr\" on \"public\".\"code_class\".\"id\" = \"public\".\"code_class_tr\".\"code_class_id\"";
    }

    @Override
    protected TableField<CodeClassRecord, Timestamp> expectedLastSyncField() {
        return ch.difty.scipamato.publ.db.tables.CodeClass.CODE_CLASS.LAST_SYNCHED;
    }

    @Test
    void makingEntity() throws SQLException {
        ResultSet rs = Mockito.mock(ResultSet.class);
        when(rs.getInt(CodeClass.CODE_CLASS.ID.getName())).thenReturn(1);
        when(rs.getString(CodeClassTr.CODE_CLASS_TR.LANG_CODE.getName())).thenReturn("lc");
        when(rs.getString(CodeClassTr.CODE_CLASS_TR.NAME.getName())).thenReturn("n");
        when(rs.getString(CodeClassTr.CODE_CLASS_TR.DESCRIPTION.getName())).thenReturn("d");
        when(rs.getInt(CodeClassTr.CODE_CLASS_TR.VERSION.getName())).thenReturn(2);
        when(rs.getTimestamp(CodeClassTr.CODE_CLASS_TR.CREATED.getName())).thenReturn(CREATED);
        when(rs.getTimestamp(CodeClassTr.CODE_CLASS_TR.LAST_MODIFIED.getName())).thenReturn(MODIFIED);

        PublicCodeClass pcc = config.makeEntity(rs);

        pcc.getCodeClassId() shouldBeEqualTo 1;
        pcc.getLangCode() shouldBeEqualTo "lc";
        pcc.getName() shouldBeEqualTo "n";
        pcc.getDescription() shouldBeEqualTo "d";
        pcc.getVersion() shouldBeEqualTo 2;
        pcc.getCreated() shouldBeEqualTo CREATED;
        pcc.getLastModified() shouldBeEqualTo MODIFIED;
        assertThat(pcc.getLastSynched()).isCloseTo("2016-12-09T06:02:13.000", 1000);

        verify{ rs.getInt(CodeClass.CODE_CLASS.ID.getName()); }
        verify{ rs.getString(CodeClassTr.CODE_CLASS_TR.LANG_CODE.getName()); }
        verify{ rs.getString(CodeClassTr.CODE_CLASS_TR.NAME.getName()); }
        verify{ rs.getString(CodeClassTr.CODE_CLASS_TR.DESCRIPTION.getName()); }
        verify{ rs.getInt(CodeClassTr.CODE_CLASS_TR.VERSION.getName()); }
        verify{ rs.getTimestamp(CodeClassTr.CODE_CLASS_TR.CREATED.getName()); }
        verify{ rs.getTimestamp(CodeClassTr.CODE_CLASS_TR.LAST_MODIFIED.getName()); }
        verify(exactly=2) { rs.wasNull(); }

        confirmVerified(rs);
    }
}

File: core/core-sync/src/test/kotlin/ch/difty/scipamato/core/sync/jobs/codeclass/CodeClassSyncConfigTest.kt
package ch.difty.scipamato.core.sync.jobs.codeclass

import ch.difty.scipamato.common.FrozenDateTimeService
import io.mockk.every
import io.mockk.mockk
import org.amshove.kluent.shouldBeEqualTo
import org.amshove.kluent.shouldBeInstanceOf
import org.amshove.kluent.shouldBeNull
import org.jooq.DSLContext
import org.jooq.SQLDialect
import org.jooq.impl.DSL
import org.jooq.tools.jdbc.MockConnection
import org.jooq.tools.jdbc.MockDataProvider
import org.junit.jupiter.api.Test
import org.springframework.batch.core.repository.JobRepository
import org.springframework.transaction.PlatformTransactionManager
import java.sql.ResultSet
import java.sql.Timestamp
import javax.sql.DataSource

internal class CodeClassSyncConfigTest {

    private val provider = mockk<MockDataProvider>()
    private val connection = MockConnection(provider)
    private val jooqCore: DSLContext = DSL.using(connection, SQLDialect.POSTGRES)
    private val jooqPublic: DSLContext = DSL.using(connection, SQLDialect.POSTGRES)
    private val coreDataSource = mockk<DataSource>()
    private val jobRepository = mockk<JobRepository>()
    private val transactionManager = mockk<PlatformTransactionManager>()
    private val dateTimeService = FrozenDateTimeService()

    private val config = CodeClassSyncConfig(
        jooqCore,
        jooqPublic,
        coreDataSource,
        jobRepository,
        transactionManager,
        dateTimeService
    )

    @Test
    fun jobName() {
        config.jobName shouldBeEqualTo "syncCodeClassJob"
    }

    @Test
    fun publicWriter() {
        config.publicWriter() shouldBeInstanceOf CodeClassItemWriter::class
    }

    @Test
    fun lastSynchedField() {
        config.lastSynchedField().toString() shouldBeEqualTo """"public"."code_class"."last_synched""""
    }

    @Test
    fun makingEntity() {
        val resultSet = mockk<ResultSet> {
            every { getInt("id") } returns 1
            every { getString("lang_code") } returns "mylangcode"
            every { getString("name") } returns "myname"
            every { getString("description") } returns "mydescription"
            every { getInt("version") } returns 2
            every { getTimestamp("created") } returns Timestamp.valueOf("2020-04-04 12:13:14")
            every { getTimestamp("last_modified") } returns Timestamp.valueOf("2020-04-04 13:14:15")
            every { wasNull() } returns false
        }

        val codeClass = config.makeEntity(resultSet)

        codeClass.codeClassId shouldBeEqualTo 1
        codeClass.langCode shouldBeEqualTo "mylangcode"
        codeClass.name shouldBeEqualTo "myname"
        codeClass.description shouldBeEqualTo "mydescription"
        codeClass.version shouldBeEqualTo 2
        codeClass.created.toString() shouldBeEqualTo "2020-04-04 12:13:14.0"
        codeClass.lastModified.toString() shouldBeEqualTo "2020-04-04 13:14:15.0"
        codeClass.lastSynched.toString() shouldBeEqualTo "2016-12-09 06:02:13.0"
    }

    @Suppress("MaxLineLength")
    @Test
    fun selectSql() {
        /* ktlint-disable max-line-length */
        config.selectSql() shouldBeEqualTo
            """select "public"."code_class"."id", "public"."code_class_tr"."lang_code", "public"."code_class_tr"."name", "public"."code_class_tr"."description", "public"."code_class_tr"."version", "public"."code_class_tr"."created", "public"."code_class_tr"."last_modified" from "public"."code_class" join "public"."code_class_tr" on "public"."code_class"."id" = "public"."code_class_tr"."code_class_id""""
        /* ktlint-enable max-line-length */
    }

    @Test
    fun pseudoFkDcs() {
        config.pseudoFkDcs.shouldBeNull()
    }
}
